### PR TITLE
[lldb][DataFormatter] VectorType: fix format for arrays with size not a power-of-2

### PR DESCRIPTION
**Pull request by Michael137**
To get the number of children for a VectorType (i.e.,
a type declared with a `vector_size`/`ext_vector_type` attribute)
LLDB previously did following calculation:
1. Get byte-size of the vector container from Clang (`getTypeInfo`).
2. Get byte-size of the element type we want to interpret the array as.
   (e.g., sometimes we want to interpret an `unsigned char vec[16]`
   as a `float32[]`).
3. `numChildren = containerSize / reinterpretedElementSize`

However, for step 1, clang will return us the *aligned* container byte-size.
So for a type such as `float __attribute__((ext_vector_type(3)))`
(which is an array of 3 4-byte floats), clang will round up the
byte-width of the array to `16`.
(see [here](https://github.com/llvm/llvm-project/blob/ab6a66dbec61654d0962f6abf6d6c5b776937584/clang/lib/AST/ASTContext.cpp#L1987-L1992))

This means that for vectors where the size isn't a power-of-2, LLDB
will miscalculate the number of elements.

**Solution**

This patch changes step 1 such that we calculate the container size
as `numElementsInSource * byteSizeOfElement`.

### CODE REVIEW

File: lldb/source/DataFormatters/VectorType.cpp [hunk 1]
```diff
@@ -169,19 +169,42 @@ static lldb::Format GetItemFormatForFormat(lldb::Format format,
...
+/// the byte-array as a `float32[]`.
+///
+/// \param[in] container_type The type of the container We
+/// are calculating the children of.
+///
+/// \param[in] element_type The type of elements we interpret
+/// container_type to contain for the purposes of calculating
+/// the number of children.
+///
+/// If size of the container is not a multiple of 'element_type'
```
Comment (medismailben) [inline lldb/source/DataFormatters/VectorType.cpp:192]: nit: You could use `\return` block here

File: lldb/source/DataFormatters/VectorType.cpp [hunk 2]
```diff
@@ -169,21 +169,49 @@ static lldb::Format GetItemFormatForFormat(lldb::Format format,
...
-      container_type.GetByteSize(exe_scope);
-  std::optional<uint64_t> element_size = element_type.GetByteSize(exe_scope);
-
-  if (container_size && element_size && *element_size) {
-    if (*container_size % *element_size)
-      return 0;
-    return *container_size / *element_size;
-  }
-  return 0;
+/// \brief Returns the number of elements stored in a container
```
Comment (adrian-prantl) [inline lldb/source/DataFormatters/VectorType.cpp:172]: We run Doxygen with autobrief, so this \brief should be redundant.

File: lldb/source/DataFormatters/VectorType.cpp [hunk 3]
```diff
@@ -169,21 +169,49 @@ static lldb::Format GetItemFormatForFormat(lldb::Format format,
...
+///
+/// \param[in] element_type The type of elements we interpret
+/// container_type to contain for the purposes of calculating
+/// the number of children.
+///
+/// If size of the container is not a multiple of 'element_type'
+/// returns 0.
+///
+/// On error, returns 0.
+static size_t CalculateNumChildren(CompilerType container_elem_type,
```
Comment (adrian-prantl) [inline lldb/source/DataFormatters/VectorType.cpp:196]: Could you, just out of principle, return optional<size_t> here even if you the use it with getValueOr(0) at the call site?

**Review by adrian-prantl — APPROVED** [review]

File: lldb/source/DataFormatters/VectorType.cpp (hunk 2 above)
Comment (Michael137) [reply lldb/source/DataFormatters/VectorType.cpp]: done

File: lldb/source/DataFormatters/VectorType.cpp (hunk 3 above)
Comment (Michael137) [reply lldb/source/DataFormatters/VectorType.cpp]: yup looks better!

**Pull request merged**